File: site-audit/parse_pages.py
#!/usr/bin/env python3
"""Parse Squarespace HTML pages into structured audit JSON."""
import json
import re
import html as html_lib
from pathlib import Path
from urllib.parse import urlparse, urlunparse

try:
    from bs4 import BeautifulSoup
except ImportError:
    import subprocess
    subprocess.check_call(["pip", "install", "beautifulsoup4", "lxml", "-q"])
    from bs4 import BeautifulSoup

AUDIT_DIR = Path(__file__).parent
PAGES = {
    "iguovbiobo": "https://www.nigeriawaterproject.org/iguovbiobo",
    "uvbevillage": "https://www.nigeriawaterproject.org/uvbevillage",
    "photos": "https://www.nigeriawaterproject.org/photos",
    "donate": "https://www.nigeriawaterproject.org/donate",
    "contact": "https://www.nigeriawaterproject.org/contact",
    "new-dropdown": "https://www.nigeriawaterproject.org/new-dropdown",
}

HOME_URL = "https://www.nigeriawaterproject.org/"


def strip_query(url: str) -> str:
    if not url:
        return url
    parsed = urlparse(url)
    return urlunparse((parsed.scheme, parsed.netloc, parsed.path, "", "", ""))


def clean_text(text: str) -> str:
    if not text:
        return ""
    text = html_lib.unescape(text)
    text = re.sub(r"\s+", " ", text).strip()
    return text


def get_main_content(soup: BeautifulSoup):
    main = soup.find("main", id="page")
    if main:
        return main
    return soup.find("div", class_="page-section") or soup.body


def extract_visible_text(main) -> dict:
    """Extract headings, paragraphs, buttons, FAQs preserving order."""
    result = {
        "headings": [],
        "paragraphs": [],
        "buttonLabels": [],
        "faqs": [],
        "allTextBlocks": [],
    }

    skip_classes = {"header-nav", "footer", "header-menu", "Mobile-bar", "sqs-announcement-bar"}

    for el in main.find_all(["h1", "h2", "h3", "h4", "h5", "h6", "p", "li", "blockquote", "figcaption"]):
        # skip nav/footer
        if any(el.find_parent(class_=re.compile(c, re.I)) for c in ["header", "footer", "Header", "Footer"]):
            parent = el
            in_nav = False
            for _ in range(10):
                parent = parent.parent
                if parent is None:
                    break
                cls = " ".join(parent.get("class", []))
                if "header-nav" in cls or "footer" in cls.lower() or "Header-nav" in cls:
                    in_nav = True
                    break
            if in_nav:
                continue

        text = clean_text(el.get_text(separator=" ", strip=True))
        if not text or len(text) < 2:
            continue

        tag = el.name
        if tag.startswith("h"):
            result["headings"].append({"level": int(tag[1]), "text": text})
        elif tag == "p":
            result["paragraphs"].append(text)
        elif tag == "li":
            # FAQ items often in accordion
            parent_acc = el.find_parent(class_=re.compile("accordion|faq", re.I))
            if parent_acc:
                result["faqs"].append(text)
            else:
                result["allTextBlocks"].append(text)
        else:
            result["allTextBlocks"].append(text)

    # Buttons
    for btn in main.find_all(["a", "button"], class_=re.compile("sqs-block-button|button|btn", re.I)):
        text = clean_text(btn.get_text(strip=True))
        if text and text not in result["buttonLabels"]:
            result["buttonLabels"].append(text)

    for btn in main.find_all("a", class_=re.compile("sqs-button-element", re.I)):
        text = clean_text(btn.get_text(strip=True))
        if text and text not in result["buttonLabels"]:
            result["buttonLabels"].append(text)

    # Accordion / FAQ blocks
    for acc in main.find_all(class_=re.compile("accordion-item|sqs-block-accordion", re.I)):
        title = acc.find(class_=re.compile("accordion-item__title|accordion-title", re.I))
        desc = acc.find(class_=re.compile("accordion-item__description|accordion-content", re.I))
        if title:
            faq_entry = {"question": clean_text(title.get_text(strip=True))}
            if desc:
                faq_entry["answer"] = clean_text(desc.get_text(separator=" ", strip=True))
            result["faqs"].append(faq_entry)

    return result


def extract_images(soup: BeautifulSoup) -> list:
    urls = set()
    for img in soup.find_all("img"):
        for attr in ["src", "data-src", "data-image", "data-image-focal-point"]:
            val = img.get(attr, "")
            if "squarespace-cdn.com" in str(val):
                urls.add(strip_query(val.split(" ")[0]))
        srcset = img.get("srcset", "")
        for part in srcset.split(","):
            u = part.strip().split(" ")[0]
            if "squarespace-cdn.com" in u:
                urls.add(strip_query(u))

    for el in soup.find_all(style=True):
        for m in re.finditer(r"url\(['\"]?(https?://images\.squarespace-cdn\.com[^)'\"]+)", el["style"]):
            urls.add(strip_query(m.group(1)))

    for el in soup.find_all(attrs={"data-image": True}):
        u = el.get("data-image", "")
        if "squarespace-cdn.com" in u:
            urls.add(strip_query(u))

    # background-image in data attributes
    html_str = str(soup)
    for m in re.finditer(r"https://images\.squarespace-cdn\.com/content/v1/[^\s\"'<>?]+", html_str):
        urls.add(strip_query(m.group(0)))

    return sorted(urls)


def extract_videos(soup: BeautifulSoup) -> list:
    videos = []
    html_str = str(soup)

    for el in soup.find_all(attrs={"data-config-video": True}):
        try:
            cfg = json.loads(el["data-config-video"])
            videos.append({"source": "data-config-video", "config": cfg})
        except json.JSONDecodeError:
            videos.append({"source": "data-config-video", "raw": el["data-config-video"]})

    for m in re.finditer(r'"alexandriaUrl"\s*:\s*"([^"]+)"', html_str):
        videos.append({"alexandriaUrl": m.group(1).replace("\\/", "/")})

    for m in re.finditer(r'"videoUrl"\s*:\s*"([^"]+)"', html_str):
        videos.append({"videoUrl": m.group(1).replace("\\/", "/")})

    for iframe in soup.find_all("iframe"):
        src = iframe.get("src", "")
        if src and ("youtube" in src or "vimeo" in src or "squarespace" in src):
            videos.append({"iframeSrc": src})

    # dedupe
    seen = set()
    unique = []
    for v in videos:
        key = json.dumps(v, sort_keys=True)
        if key not in seen:
            seen.add(key)
            unique.append(v)
    return unique


def extract_layout(soup: BeautifulSoup) -> dict:
    sections = []
    main = get_main_content(soup)
    for i, sec in enumerate(main.find_all("section", class_=re.compile("page-section", re.I))):
        info = {"index": i}
        classes = sec.get("class", [])
        info["classes"] = classes

        # background color from data attributes or inline style
        bg = sec.get("data-section-theme") or sec.get("data-section-id")
        if bg:
            info["dataSectionTheme"] = bg

        style = sec.get("style", "")
        bg_match = re.search(r"background(?:-color)?:\s*([^;]+)", style, re.I)
        if bg_match:
            info["backgroundColor"] = bg_match.group(1).strip()

        # Check for background image wrapper
        bg_div = sec.find(class_=re.compile("section-background|background-image", re.I))
        if bg_div:
            bg_style = bg_div.get("style", "")
            m = re.search(r"background(?:-color)?:\s*([^;]+)", bg_style, re.I)
            if m:
                info["backgroundColor"] = m.group(1).strip()

        # Column layout
        content = sec.find(class_=re.compile("content-wrapper|section-content", re.I))
        if content:
            cols = content.find_all(class_=re.compile("sqs-col|fluid-engine", re.I))
            if cols:
                info["hasColumns"] = True
                info["columnCount"] = len(cols)

        layout_type = None
        for c in classes:
            if "layout-engine" in c:
                layout_type = c
        if layout_type:
            info["layoutType"] = layout_type

        sections.append(info)

    return {
        "sectionCount": len(sections),
        "sections": sections[:20],
        "maxPageWidth": "1400px",
        "pagePadding": "4vw",
        "transparentHeader": True,
    }


def extract_forms(soup: BeautifulSoup) -> list:
    forms = []
    for form in soup.find_all("form"):
        fields = []
        for inp in form.find_all(["input", "textarea", "select"]):
            ftype = inp.name
            field = {
                "tag": ftype,
                "type": inp.get("type", ftype),
                "name": inp.get("name", ""),
                "id": inp.get("id", ""),
                "placeholder": inp.get("placeholder", ""),
                "required": inp.has_attr("required"),
            }
            label = form.find("label", attrs={"for": inp.get("id")})
            if label:
                field["label"] = clean_text(label.get_text(strip=True))
            fields.append(field)

        # Squarespace form blocks use data attributes
        form_block = form.find_parent(class_=re.compile("form-block", re.I))
        submit = form.find(["button", "input"], type="submit")
        forms.append({
            "action": form.get("action", ""),
            "method": form.get("method", "GET"),
            "fields": fields,
            "submitLabel": clean_text(submit.get_text(strip=True)) if submit else "Submit",
        })
    return forms


def extract_external_links(soup: BeautifulSoup) -> list:
    links = []
    domain = "nigeriawaterproject.org"
    for a in soup.find_all("a", href=True):
        href = a["href"]
        text = clean_text(a.get_text(strip=True))
        if not href.startswith("http"):
            continue
        if domain in href:
            continue
        links.append({"href": href, "text": text})

    # dedupe by href
    seen = set()
    unique = []
    for l in links:
        if l["href"] not in seen:
            seen.add(l["href"])
            unique.append(l)
    return unique


def extract_nav_dropdown(soup: BeautifulSoup) -> list:
    """Extract Projects dropdown submenu items."""
    items = []
    for a in soup.find_all("a", href=True):
        href = a["href"]
        text = clean_text(a.get_text(strip=True))
        # submenu items are village pages
        if any(v in href for v in ["iguovbiobo", "uvbevillage", "ebueneki", "urhokuosa", "new-dropdown"]):
            if text and len(text) > 2:
                items.append({"label": text, "href": href})

    # Also look for folder dropdown structure
    for folder in soup.find_all(class_=re.compile("header-nav-folder|folder-item", re.I)):
        for a in folder.find_all("a", href=True):
            text = clean_text(a.get_text(strip=True))
            href = a["href"]
            if text:
                items.append({"label": text, "href": href})

    # dedupe
    seen = set()
    unique = []
    for item in items:
        key = (item["label"], item["href"])
        if key not in seen:
            seen.add(key)
            unique.append(item)
    return unique


def parse_page(slug: str, url: str) -> dict:
    html_path = AUDIT_DIR / f"{slug}.html"
    if not html_path.exists():
        return {"url": url, "error": f"HTML file not found: {html_path}"}

    html = html_path.read_text(encoding="utf-8", errors="replace")
    soup = BeautifulSoup(html, "lxml")

    title_tag = soup.find("title")
    seo_title = clean_text(title_tag.get_text()) if title_tag else ""

    main = get_main_content(soup)

    return {
        "url": url,
        "slug": slug,
        "seoTitle": seo_title,
        "visibleText": extract_visible_text(main),
        "imageUrls": extract_images(soup),
        "videos": extract_videos(soup),
        "layout": extract_layout(soup),
        "forms": extract_forms(soup),
        "externalLinks": extract_external_links(soup),
    }


def main():
    output = {
        "auditDate": "2026-07-12",
        "source": "HTML crawl (Squarespace server-rendered)",
        "projectsDropdown": [],
        "pages": {},
    }

    home_path = AUDIT_DIR / "home.html"
    if home_path.exists():
        home_soup = BeautifulSoup(home_path.read_text(encoding="utf-8", errors="replace"), "lxml")
        output["projectsDropdown"] = extract_nav_dropdown(home_soup)

    for slug, url in PAGES.items():
        output["pages"][slug] = parse_page(slug, url)

    out_path = AUDIT_DIR / "page-content.json"
    out_path.write_text(json.dumps(output, indent=2, ensure_ascii=False), encoding="utf-8")
    print(f"Wrote {out_path}")
    print(f"Pages: {len(output['pages'])}")
    print(f"Projects dropdown items: {len(output['projectsDropdown'])}")


if __name__ == "__main__":
    main()

File: site-audit/parse_pages.js
const fs = require('fs');
const path = require('path');

const AUDIT_DIR = __dirname;
const PAGES = {
  iguovbiobo: 'https://www.nigeriawaterproject.org/iguovbiobo',
  uvbevillage: 'https://www.nigeriawaterproject.org/uvbevillage',
  photos: 'https://www.nigeriawaterproject.org/photos',
  donate: 'https://www.nigeriawaterproject.org/donate',
  contact: 'https://www.nigeriawaterproject.org/contact',
  'new-dropdown': 'https://www.nigeriawaterproject.org/new-dropdown',
};

const THEME_COLORS = {
  white: 'rgb(251, 251, 251)',
  black: 'rgb(52, 47, 47)',
  light: 'rgb(232, 232, 232)',
  'dark-bold': 'rgb(52, 47, 47)',
  dark: 'rgb(79, 76, 76)',
  'bright-inverse': 'rgb(251, 251, 251)',
  '': 'transparent/default',
};

function stripQuery(url) {
  return url.split('?')[0].replace(/&quot;.*$/, '').replace(/\\$/, '');
}

function decodeHtml(str) {
  return str
    .replace(/&amp;/g, '&')
    .replace(/&lt;/g, '<')
    .replace(/&gt;/g, '>')
    .replace(/&quot;/g, '"')
    .replace(/&#39;/g, "'")
    .replace(/&mdash;/g, '—')
    .replace(/&nbsp;/g, ' ')
    .replace(/<br\s*\/?>/gi, '\n')
    .replace(/\s+/g, (m, offset, s) => (s[offset - 1] === '\n' || s[offset + m.length] === '\n' ? ' ' : ' '))
    .trim();
}

function extractMainContent(html) {
  const mainMatch = html.match(/<main[^>]*id="page"[^>]*>([\s\S]*?)<\/main>/i);
  if (mainMatch) {
    const footerIdx = mainMatch[1].lastIndexOf('fe-67137859ccf3f0398022af6d');
    if (footerIdx > 0) return mainMatch[1].slice(0, footerIdx);
    return mainMatch[1];
  }
  return html;
}

function extractTitle(html) {
  const m = html.match(/<title>([^<]*)<\/title>/i);
  return m ? decodeHtml(m[1]) : '';
}

function extractHeadings(content) {
  const headings = [];
  const re = /<(h[1-6])[^>]*>([\s\S]*?)<\/\1>/gi;
  let m;
  while ((m = re.exec(content)) !== null) {
    let text = m[2].replace(/<br\s*\/?>/gi, '\n').replace(/<[^>]+>/g, '');
    text = decodeHtml(text);
    if (text && !/^(NIGERIA WATER PROJECT|Location|Contact)$/i.test(text)) {
      headings.push({ level: parseInt(m[1][1]), text });
    }
  }
  return headings;
}

function extractParagraphs(content) {
  const paragraphs = [];
  const re = /<p[^>]*>([\s\S]*?)<\/p>/gi;
  let m;
  while ((m = re.exec(content)) !== null) {
    const text = decodeHtml(m[1].replace(/<[^>]+>/g, ''));
    if (text && text.length > 1 && !/^(STANFORD|TAMPA|nigeriawaterproject|mudiar|ogbereuben)/i.test(text)) {
      paragraphs.push(text);
    }
  }
  return paragraphs;
}

function extractButtons(content) {
  const labels = new Set();
  const re = /class="[^"]*sqs-block-button[^"]*"[\s\S]*?<a[^>]*>([\s\S]*?)<\/a>/gi;
  let m;
  while ((m = re.exec(content)) !== null) {
    const text = decodeHtml(m[1].replace(/<[^>]+>/g, ''));
    if (text) labels.add(text);
  }
  const re2 = /class="dbox-donation-button"[^>]*>[\s\S]*?<span[^>]*>([^<]*)<\/span>/gi;
  while ((m = re2.exec(content)) !== null) {
    if (m[1]) labels.add(decodeHtml(m[1]));
  }
  return [...labels];
}

function extractFaqs(html) {
  const faqs = [];
  const re = /class="accordion-item__title">([^<]*)<[\s\S]*?class="[\s\S]*?accordion-item__description[\s\S]*?>([\s\S]*?)<\/div>\s*<\/div>/gi;
  let m;
  while ((m = re.exec(html)) !== null) {
    faqs.push({
      question: decodeHtml(m[1]),
      answer: decodeHtml(m[2].replace(/<[^>]+>/g, '')),
    });
  }
  return faqs;
}

function extractImages(html) {
  const urls = new Set();
  const re = /https:\/\/images\.squarespace-cdn\.com\/content\/v1\/[^\s"'<>?\\]+/g;
  let m;
  while ((m = re.exec(html)) !== null) {
    const clean = stripQuery(m[0]);
    if (!clean.endsWith('Logo.png') && !clean.endsWith('favicon.ico')) {
      urls.add(clean);
    }
  }
  return [...urls].sort();
}

function extractVideos(html) {
  const videos = [];
  const re = /data-config-video="([^"]+)"/g;
  let m;
  while ((m = re.exec(html)) !== null) {
    const decoded = m[1].replace(/&quot;/g, '"').replace(/&amp;/g, '&');
    try {
      videos.push({ source: 'data-config-video', config: JSON.parse(decoded) });
    } catch {
      videos.push({ source: 'data-config-video', raw: decoded });
    }
  }
  return videos;
}

function extractLayout(html) {
  const sections = [];
  const re = /<section[^>]*data-section-id="([^"]*)"[^>]*data-section-theme="([^"]*)"[^>]*>/gi;
  let m;
  let i = 0;
  while ((m = re.exec(html)) !== null) {
    const theme = m[2];
    const sec = {
      index: i++,
      sectionId: m[1],
      sectionTheme: theme || 'default',
      backgroundColorRgb: THEME_COLORS[theme] || THEME_COLORS[''],
    };
    const chunk = html.slice(m.index, m.index + 2000);
    const overlay = chunk.match(/section-background-overlay[^>]*style="opacity:\s*([^"]+)"/);
    if (overlay) sec.backgroundOverlayOpacity = overlay[1];
    const hasFluid = chunk.includes('fluid-engine');
    if (hasFluid) sec.layout = 'fluid-engine (multi-column grid)';
    sections.push(sec);
  }
  return {
    sectionCount: sections.length,
    sections,
    maxPageWidth: '1400px',
    pagePadding: '4vw',
    transparentHeader: html.includes('tweak-transparent-header'),
    headerWidth: 'Inset',
    fonts: ['Anton (headings)', 'Epilogue (body)'],
  };
}

function extractForms(html) {
  const forms = [];
  const ctxRe = /id="form-context-[^"]*"[^>]*>\s*(\{[\s\S]*?\})\s*<\/script>/gi;
  let m;
  while ((m = ctxRe.exec(html)) !== null) {
    try {
      const ctx = JSON.parse(m[1]);
      const fields = (ctx.form || ctx.fields || []).map((f) => ({
        id: f.id,
        type: f.type,
        title: f.title || f.label || '',
        required: f.required || false,
        options: f.options || undefined,
      }));
      forms.push({
        formId: ctx.formId || ctx.collectionId,
        fields,
        submitLabel: 'Submit',
        storageNote: ctx.useLightbox ? 'lightbox form' : 'inline form',
      });
    } catch {}
  }

  const formRe = /<form[^>]*>([\s\S]*?)<\/form>/gi;
  while ((m = formRe.exec(html)) !== null) {
    const fields = [];
    const fieldRe = /<(input|textarea|select)[^>]*>/gi;
    let fm;
    while ((fm = fieldRe.exec(m[1])) !== null) {
      const tag = fm[0];
      fields.push({
        tag: fm[1],
        type: tag.match(/type="([^"]*)"/)?.[1] || fm[1],
        name: tag.match(/name="([^"]*)"/)?.[1] || '',
        placeholder: tag.match(/placeholder="([^"]*)"/)?.[1] || '',
        required: tag.includes('required'),
      });
    }
    if (fields.length) forms.push({ fields, submitLabel: 'Submit' });
  }
  return forms;
}

function extractExternalLinks(html) {
  const links = [];
  const seen = new Set();
  const re = /<a[^>]*href="(https?:\/\/[^"]+)"[^>]*>([\s\S]*?)<\/a>/gi;
  let m;
  while ((m = re.exec(html)) !== null) {
    const href = m[1];
    if (href.includes('nigeriawaterproject.org') || href.includes('squarespace.com')) continue;
    if (seen.has(href)) continue;
    seen.add(href);
    const text = decodeHtml(m[2].replace(/<[^>]+>/g, '')) || href;
    links.push({ href, text });
  }
  return links;
}

function extractProjectsDropdown(html) {
  const items = [];
  const re = /<a\s+href="(\/[^"]+)"\s*>([^<]*)<\/a>/gi;
  const folderContent = html.match(/header-nav-folder-content[^>]*id="projects"[^>]*>([\s\S]*?)<\/div>\s*<\/div>/i);
  if (folderContent) {
    let m;
    while ((m = re.exec(folderContent[1])) !== null) {
      items.push({ label: decodeHtml(m[2]), href: m[1] });
    }
  }
  return items;
}

function extractScrollingText(html) {
  const items = [];
  const re = /"sectionTitle"\s*:\s*"([^"]+)"/g;
  let m;
  while ((m = re.exec(html)) !== null) {
    const decoded = m[1].replace(/\\"/g, '"').replace(/\\n/g, '\n').replace(/\\u003c/g, '<').replace(/\\u003e/g, '>');
    const text = decodeHtml(decoded.replace(/<[^>]+>/g, ''));
    if (text) items.push(text);
  }
  return items;
}

function parsePage(slug, url) {
  const htmlPath = path.join(AUDIT_DIR, `${slug}.html`);
  const html = fs.readFileSync(htmlPath, 'utf8');
  const content = extractMainContent(html);

  const result = {
    url,
    slug,
    seoTitle: extractTitle(html),
    visibleText: {
      headings: extractHeadings(content),
      paragraphs: extractParagraphs(content),
      buttonLabels: extractButtons(html),
      scrollingText: extractScrollingText(html),
      faqs: extractFaqs(html),
    },
    imageUrls: extractImages(html),
    videos: extractVideos(html),
    layout: extractLayout(html),
    forms: extractForms(html),
    externalLinks: extractExternalLinks(html),
  };

  return result;
}

const homeHtml = fs.readFileSync(path.join(AUDIT_DIR, 'home.html'), 'utf8');

const output = {
  auditDate: '2026-07-12',
  source: 'HTML crawl (Squarespace server-rendered; browser MCP unavailable in this session)',
  themeColorMap: THEME_COLORS,
  projectsDropdown: extractProjectsDropdown(homeHtml),
  pages: {},
};

for (const [slug, url] of Object.entries(PAGES)) {
  output.pages[slug] = parsePage(slug, url);
}

const outPath = path.join(AUDIT_DIR, 'page-content.json');
fs.writeFileSync(outPath, JSON.stringify(output, null, 2), 'utf8');
console.log('Done');
